### PR TITLE
AIT-402 | Including Matt Hager's threaded runtime. Restructured that run...

### PR DESCRIPTION
**Pull request by robdouglas**
...time's project organization a bit to be more in line with the streams-runtime-local

### CODE REVIEW

**Comment by smashew** [comment]
:+1:

File: streams-runtimes/streams-runtime-threaded/pom.xml [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-core</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>org.apache.streams</groupId>
+            <artifactId>streams-core</artifactId>
+            <version>${project.version}</version>
```
Comment (mfranklin) [inline streams-runtimes/streams-runtime-threaded/pom.xml]: versions should be inherited from the defined dependency in parent dependency management

File: streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/builders/ThreadedStreamBuilder.java [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+    public ThreadedStreamBuilder(Queue<StreamsDatum> queue, Map<String, Object> streamConfig, ThreadingController threadingController) {
+        this.queue = queue;
+        this.providers = new LinkedHashMap<String, StreamComponent>();
+        this.components = new LinkedHashMap<String, StreamComponent>();
+        this.streamConfig = streamConfig;
+        this.threadingController = threadingController;
+    }
+
+    public List<StreamsGraphElement> getGraphElements() {
+        if(this.graphElements.size() == 0)
```
Comment (sieverssj) [inline streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/builders/ThreadedStreamBuilder.java]: Style nit: Wrap in curly braces please.

File: streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/builders/ThreadedStreamBuilder.java [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+        synchronized (ThreadedStreamBuilder.class) {
+            return Collections.unmodifiableList(CURRENTLY_EXECUTING);
+        }
+    }
+
+    private void buildGraphElements() {
+        this.graphElements.clear();
+        for(StreamsTask p : this.tasks.values()) {
+            if(p instanceof StreamsProviderTask)
+                appendAndFollow(p, this.graphElements);
```
Comment (sieverssj) [inline streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/builders/ThreadedStreamBuilder.java]: Style nit: Wrap in curly braces please.

File: streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/builders/ThreadedStreamBuilder.java [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+        }
+    }
+
+    private void appendAndFollow(StreamsTask t, List<StreamsGraphElement> elems) {
+        for(StreamsTask c : t.getChildren()) {
+
+            String type;
+            if(t.getClass().equals(StreamsProviderTask.class))
```
Comment (sieverssj) [inline streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/builders/ThreadedStreamBuilder.java]: Style nit: Wrap in curly braces please.

File: streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/builders/ThreadedStreamBuilder.java [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+        return this;
+    }
+
+    public ThreadedStreamBuilder addEventHandler(StreamBuilderEventHandler eventHandler) {
+        this.eventHandlers.add(eventHandler);
+        return this;
+    }
+
+    public ThreadedStreamBuilder removeEventHandler(StreamBuilderEventHandler eventHandler) {
+        if(this.eventHandlers.contains(eventHandler))
```
Comment (sieverssj) [inline streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/builders/ThreadedStreamBuilder.java]: Style nit: Wrap in curly braces please.

File: streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/builders/ThreadedStreamBuilder.java [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+            this.eventHandlers.remove(eventHandler);
+        return this;
+    }
+
+    public final Map<String, StatusCounts> getUpdateCounts() {
+        final Map<String, StatusCounts> updateMap = new HashMap<String, StatusCounts>();
+
+        for (final String k : tasks.keySet())
```
Comment (sieverssj) [inline streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/builders/ThreadedStreamBuilder.java]: Style nit: Wrap in curly braces please.

File: streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/builders/ThreadedStreamBuilder.java [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+        createTasks();
+
+        // if anyone would like to listen in to progress events
+        // let them do that
+        final TimerTask updateTask = getUpdateCounts() == null || getUpdateCounts().keySet().size() == 0 ? null : new TimerTask() {
+            public void run() {
+
+                final Map<String, StatusCounts> updateMap = getUpdateCounts();
+
+                for(String k : updateMap.keySet())
```
Comment (sieverssj) [inline streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/builders/ThreadedStreamBuilder.java]: Style nit: Wrap in curly braces please.

File: streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/builders/ThreadedStreamBuilder.java [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+                            g.setValue((int)updateMap.get(k).getWorking());
+
+
+                if (eventHandlers.size() > 0) {
+                    for (final StreamBuilderEventHandler eventHandler : eventHandlers) {
+                        try {
+                            try {
+                                eventHandler.update(updateMap, getGraphElements());
+                            } catch(Throwable e) {
+                                    /* */
```
Comment (sieverssj) [inline streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/builders/ThreadedStreamBuilder.java]: Please log these exceptions

File: streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/builders/ThreadedStreamBuilder.java [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+            }
+            catch (Throwable omgE) {
+                LOGGER.error("Unexpected Error: {}", omgE);
+            }
+        } finally {
+            synchronized (ThreadedStreamBuilder.class) {
+                CURRENTLY_EXECUTING.remove(this);
+            }
+            // kill the updateTask
+            if(updateTask != null)
```
Comment (sieverssj) [inline streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/builders/ThreadedStreamBuilder.java]: Style nit: Wrap in curly braces please.

**Comment by sieverssj** [comment]
I started documenting places where there were missing curly braces on conditionals and for loops.  I stopped after I realized there were many more places that lacked curly braces.  2 characters is a small price to pay to mitigate all of the maintainability issues that lacking them causes.

File: streams-runtimes/streams-runtime-threaded/pom.xml [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        <dependency>
+            <groupId>ch.qos.logback</groupId>
+            <artifactId>logback-classic</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>log4j-over-slf4j</artifactId>
+        </dependency>
+        <!-- This ensures slf4j-log4j12 is not packaged in implementations -->
+        <dependency>
```
Comment (sieverssj) [inline streams-runtimes/streams-runtime-threaded/pom.xml]: I'm not sure why this is included at build-time but not at run-time.

File: streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/controller/ThreadingController.java [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+
+    private volatile double lastCPUObservation = 0.0;
+
+    private static final long SCALE_CHECK = 2000;
+    private Double scaleThreshold = .85;
+    private ThreadingControllerCPUObserver threadingControllerCPUObserver = new DefaultThreadingControllerCPUObserver();
+
+    private static final Integer NUM_PROCESSORS = Runtime.getRuntime().availableProcessors();
+
+    public static final ThreadingController INSTANCE_LOW_PRIORITY = new ThreadingController("Apache Streams [low]", NUM_PROCESSORS, NUM_PROCESSORS * 2, Thread.NORM_PRIORITY - 2);
```
Comment (sieverssj) [inline streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/controller/ThreadingController.java]: Any reason these aren't being initialized lazily in the static getters?

File: streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/controller/ThreadingController.java [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+    public String getName() {
+        return name;
+    }
+
+    /**
+     * The current number of threads in the core pool
+     * @return
+     * Integer representing the number of threads in the core pool.
+     */
+    public Integer getNumThreads()              { return this.numThreads.get(); }
```
Comment (sieverssj) [inline streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/controller/ThreadingController.java]: Please check the formatting in this class.  These are not the conventions used elsewhere in this project.

File: streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/controller/ThreadingController.java [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+            }
+
+            this.numThreads.set(newThreadCount);
+
+            // wait
+            while (this.workingNow.get() >= this.numThreads.get()) {
+                try {
+                    this.lock.await();
+                } catch (InterruptedException e) {
+                    /* no operation */
```
Comment (sieverssj) [inline streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/controller/ThreadingController.java]: At least log this.

File: streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/controller/ThreadingController.java [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+                this.lastWorked.set(new Date().getTime());
+                this.numberOfObservations.set(0);
+                this.sumOfObservations.set(0);
+            }
+
+            while (this.workingNow.get() >= this.numThreads.get()) {
+                try {
+                    this.lock.await();
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
```
Comment (sieverssj) [inline streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/controller/ThreadingController.java]: Log please.

**Comment by sieverssj** [comment]
Please also ensure that the Apache license has been applied to all files.

File: streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/tasks/StatusCounts.java [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+
+    private String id;
+    private String type;
+    private long working;
+    private long success;
+    private long failed;
+    private long timeSpentSuccess;
+    private long timeSpentFailure;
+
+    public String getId()                   { return id; }
```
Comment (sieverssj) [inline streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/tasks/StatusCounts.java]: Please check the formatting in this class as it doesn't match the conventions in the rest of the project.

File: streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/controller/ThreadingController.java (hunk 11 above)
Comment (robdouglas) [inline streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/controller/ThreadingController.java]: It appears that the getters are not actually getting used in this runtime. Instead, these attributes are publicly accessible so they are just being accessed as attributes of the ThreadedStreamBuilder class (which would necessitate initialization here or in the constructor)

**Comment by robdouglas** [comment]
Responded to code review feedback

File: streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/controller/ThreadingController.java (hunk 11 above)
Comment (sieverssj) [inline streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/controller/ThreadingController.java]: That's an anti-pattern to me.  I would rather see these be private and third-parties access them through the static getters that do lazy initialization.

**Comment by smashew** [comment]
:+1:

File: streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/controller/ThreadingController.java (hunk 11 above)
Comment (smashew) [inline streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/controller/ThreadingController.java]: they are private atomic values but the reported values are the primitives letting the user know what is happening.

Comment (mfranklin) [inline streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/controller/ThreadingController.java]: public static final ThreadingController INSTANCE = new ThreadingController("Apache Streams [default]", NUM_PROCESSORS, NUM_PROCESSORS * 5, Thread.NORM_PRIORITY);
    public static final ThreadingController INSTANCE_HIGH_PRIORITY = new ThreadingController("Apache Streams [high]", NUM_PROCESSORS, NUM_PROCESSORS * 7, Thread.NORM_PRIORITY + 2);

This is what stanton is referring to.  These should be private and only accessed via a static getter.

**Comment by robdouglas** [comment]
Responded to code review feedback

**Pull request merged**